libraries/AP_RCProtocol/AP_RCProtocol_CRSF.cpp

```diff
@@ -212,76 +212,126 @@ uint16_t AP_RCProtocol_CRSF::get_link_rate(ProtocolType protocol) const {
+    // check for long frame gaps
+    // we took too long decoding, start again - the RX will only send complete frames so this is unlikely to fail,
+    // however thread scheduling can introduce longer delays even when the data has been received
+    if (_frame_ofs > 0 && (now - _start_frame_time_us) > CRSF_FRAME_TIMEOUT_US) {
         _frame_ofs = 0;
     }
 
     // start of a new frame
     if (_frame_ofs == 0) {
-        _start_frame_time_us = timestamp_us;
+        _start_frame_time_us = now;
+    }
+    
+    _frame_bytes[_frame_ofs++] = byte;
+    
+    if (!check_frame(now)) {
+        skip_to_next_frame(now);
     }
+}
 
-    add_to_buffer(_frame_ofs++, byte);
+// check if a frame is valid. Return false if the frame is definitely
+// invalid. Return true if we need more bytes
+bool AP_RCProtocol_CRSF::check_frame(uint32_t timestamp_us)
+{
+    // overflow check
+    if (_frame_ofs >= sizeof(_frame)) {
+        return false;
+    }
 
     // need a header to get the length
     if (_frame_ofs < CRSF_HEADER_TYPE_LEN) {
-        return;
+        return true;
     }
 
     if (_frame.device_address != DeviceAddress::CRSF_ADDRESS_FLIGHT_CONTROLLER) {
-        return;
-    }
-
-    // parse the length
-    if (_frame_ofs == CRSF_HEADER_TYPE_LEN) {
-        _frame_crc = crc8_dvb_s2(0, _frame.type);
-        // check for garbage frame
-        if (_frame.length > CRSF_FRAME_PAYLOAD_MAX) {
-            _frame_ofs = 0;
-        }
-        return;
-    }
-
-    // update crc
-    if (_frame_ofs < _frame.length + CRSF_HEADER_LEN) {
-        _frame_crc = crc8_dvb_s2(_frame_crc, byte);
+        return false;
     }
 
-    // overflow check
-    if (_frame_ofs > _frame.length + CRSF_HEADER_LEN) {
-        _frame_ofs = 0;
-        return;
+    // check validity of the length byte if we have received it
+    if (_frame_ofs >= CRSF_HEADER_TYPE_LEN &&
+        _frame.length > CRSF_FRAME_PAYLOAD_MAX) {
+        return false;
     }
 
     // decode whatever we got and expect
-    if (_frame_ofs == _frame.length + CRSF_HEADER_LEN) {
-        log_data(AP_RCProtocol::CRSF, timestamp_us, (const uint8_t*)&_frame, _frame_ofs - CRSF_HEADER_LEN);
+    if (_frame_ofs >= _frame.length + CRSF_HEADER_LEN) {
+        const uint8_t crc = crc8_dvb_s2_update(0, &_frame_bytes[CRSF_HEADER_LEN], _frame.length - 1);
 
-        // we consumed the partial frame, reset
-        _frame_ofs = 0;
+        //debug("check_frame(0x%x, 0x%x)", _frame.device_address, _frame.length);
 
-        // bad CRC (payload start is +1 from frame start, so need to subtract that from frame length to get index)
-        if (_frame_crc != _frame.payload[_frame.length - 2]) {
-            return;
+        if (crc != _frame.payload[_frame.length - 2]) {
+            return false;
         }
 
+        log_data(AP_RCProtocol::CRSF, timestamp_us, (const uint8_t*)&_frame, _frame.length + CRSF_HEADER_LEN);
+
+        // we consumed the frame
+        _frame_ofs -= _frame.length + CRSF_HEADER_LEN;
```

Comment:
the _frame_ofs change should be after decode_crsf_packet() and then followed by a memmove(), or we could lose bytes